File: api/grpc/src/services/media.rs
use grpc::{ServerRequestSingle, ServerResponseUnarySink};

use mizer_api::handlers::MediaHandler;
use mizer_api::models::*;

use crate::protos::MediaApi;

impl MediaApi for MediaHandler {
    fn create_tag(
        &self,
        req: ServerRequestSingle<CreateMediaTag>,
        resp: ServerResponseUnarySink<MediaTag>,
    ) -> grpc::Result<()> {
        let handler = self.clone();
        let name = req.message.name.clone();
        req.loop_handle().spawn(async move {
            let tag = handler.create_tag(name).await.unwrap();

            resp.finish(tag)
        });

        Ok(())
    }

    fn get_tags_with_media(
        &self,
        req: ServerRequestSingle<GetMediaTags>,
        resp: ServerResponseUnarySink<GroupedMediaFiles>,
    ) -> grpc::Result<()> {
        let handler = self.clone();
        req.loop_handle().spawn(async move {
            let files = handler.get_tags_with_media().await.unwrap();

            resp.finish(files)
        });

        Ok(())
    }

    fn get_media(
        &self,
        req: ServerRequestSingle<GetMediaRequest>,
        resp: ServerResponseUnarySink<MediaFiles>,
    ) -> grpc::Result<()> {
        let handler = self.clone();
        req.loop_handle().spawn(async move {
            let files = handler.get_media().await.unwrap();

            resp.finish(files)
        });

        Ok(())
    }
}

File: ui/lib/api/grpc/media.dart
import 'package:grpc/grpc.dart';
import 'package:mizer/api/contracts/media.dart';
import 'package:mizer/protos/media.pbgrpc.dart';

class MediaGrpcApi implements MediaApi {
  final MediaApiClient client;

  MediaGrpcApi(ClientChannel channel) : client = MediaApiClient(channel);

  @override
  Future<MediaTag> createTag(String name) {
    return this.client.createTag(CreateMediaTag(name: name));
  }

  @override
  Future<MediaFiles> getMedia() {
    return this.client.getMedia(GetMediaRequest());
  }

  @override
  Future<GroupedMediaFiles> getTagsWithMedia() {
    return this.client.getTagsWithMedia(GetMediaTags());
  }
}
